cmd/generic-handlers.go

```diff
@@ -469,3 +477,90 @@ func (h pathValidityHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	}
 	h.handler.ServeHTTP(w, r)
 }
+
+// logginHandler logs all HTTP requests and responses
+// and write the log to a io.Writer object.
+type loggingHandler struct {
+	writer    io.Writer
+	writeLock *sync.Mutex
+	handler   http.Handler
+}
+
+func (h loggingHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+
+	const timeFormat = "2006-01-02 15:04:05 -0700"
+	const recordLen = 32 * 1024
+
+	// Generate short random request ID
+	reqID := fmt.Sprintf("%X", time.Now().UnixNano()%1e8)
+
+	reqTime := time.Now().Format(timeFormat)
+
+	// Setup a http request body recorder
+	reqBodyRecorder := &recordReader{Reader: r.Body, recLen: recordLen}
+	r.Body = ioutil.NopCloser(reqBodyRecorder)
+
+	// Setup a http response body recorder
+	rec := httptest.NewRecorder()
+	respBodyRecorder := &recordResponseWriter{ResponseWriter: rec, recLen: recordLen}
+
+	h.handler.ServeHTTP(respBodyRecorder, r)
+	respTime := time.Now().Format(timeFormat)
+
+	// Copy the captured response headers to the new response
+	for k, v := range rec.Header() {
+		w.Header()[k] = v
+	}
+	w.WriteHeader(rec.Code)
+	// Copy the captured response body to the ew response
+	_, wErr := rec.Body.WriteTo(w)
```

Comment:
```
+	rec := httptest.NewRecorder()
+	respBodyRecorder := &recordResponseWriter{ResponseWriter: rec, recLen: recordLen}
```

this would mean that PutObjectHandler writes to `rec` and hence all the contents will have to be stored in memory, right?

> your implementation alternative is wonderful, but we will need to include another dependency (forward), are we okay with that ?

`httpdump` is not written to be included as a dependency, I was just pointing out to an alternate implementation if httptest.NewRecorder() does not solve memory problem